components/match2/wikis/clashroyale/match_group_input_custom.lua

```diff
@@ -0,0 +1,702 @@
+
+	-- get participants data for the map + get map mode
+	map = CustomMatchGroupInput._processPlayerMapData(map, match)
+
+	-- set sumscore to 0 if it isn't a number
+	if Logic.isEmpty(match.opponent1.sumscore) then
+		match.opponent1.sumscore = 0
+	end
+	if Logic.isEmpty(match.opponent2.sumscore) then
+		match.opponent2.sumscore = 0
+	end
+
+	--adjust sumscore for winner opponent
+	if (tonumber(map.winner) or 0) > 0 then
+		match['opponent' .. map.winner].sumscore =
+			match['opponent' .. map.winner].sumscore + 1
+	end
+
+	match['map' .. mapIndex] = map
+
+	return match, subGroupIndex
+end
+
+---@param map table
+---@return table
+function CustomMatchGroupInput._mapWinnerProcessing(map)
+	if map.winner == 'skip' then
+		map.scores = {NO_SCORE, NO_SCORE}
+		map.resulttype = 'np'
+
+		return map
+	end
+
+	map.scores = {}
+	local hasManualScores = false
+
+	local scores = Array.map(Array.range(1, MAX_NUM_OPPONENTS), function(opponentIndex)
+		local score = map['score' .. opponentIndex]
+		map.scores[opponentIndex] = tonumber(score) or NO_SCORE
+
+		if String.isEmpty(score) then
+			hasManualScores = true
+		end
+
+		return Table.includes(ALLOWED_STATUSES, string.upper(score or ''))
+			and score:upper()
+			or map.scores[opponentIndex]
+	end)
+
+	if not hasManualScores then
+		local winnerInput = tonumber(map.winner)
+		if winnerInput == 1 then
+			map.scores = {1, 0}
+		elseif winnerInput == 2 then
+			map.scores = {0, 1}
+		end
+
+		return map
+	end
+
+	walkoverProcessing.walkover(map, scores)
+
+	return map
+end
+
+---@param map table
+---@param match table
+---@return table
+function CustomMatchGroupInput._processPlayerMapData(map, match)
+	local participants = {}
+
+	for opponentIndex = 1, MAX_NUM_OPPONENTS do
+		local opponent = match['opponent' .. opponentIndex]
+		if Opponent.typeIsParty(opponent.type) then
+			CustomMatchGroupInput._processDefaultPlayerMapData(
+				opponent.match2players or {},
+				opponentIndex,
+				map,
+				participants
+			)
+		elseif opponent.type == Opponent.team then
+			CustomMatchGroupInput._processTeamPlayerMapData(
+				opponent.match2players or {},
+				opponentIndex,
+				map,
+				participants
+			)
+		end
+	end
+
+	map.mode = Opponent.toMode(match.opponent1.type, match.opponent2.type)
+
+	map.participants = participants
+
+	return map
+end
+
+---@param players table
+---@param opponentIndex integer
+---@param map table
+---@param participants table
+function CustomMatchGroupInput._processDefaultPlayerMapData(players, opponentIndex, map, participants)
+	for playerIndex = 1, #players do
+		participants[opponentIndex .. '_' .. playerIndex] = {
+			played = true,
+			cards = CustomMatchGroupInput._readCards(map['t' .. opponentIndex .. 'p' .. playerIndex .. 'c']),
+		}
+	end
+end
+
+---@param players table
+---@param opponentIndex integer
+---@param map table
+---@param participants table
+---@return integer
+function CustomMatchGroupInput._processTeamPlayerMapData(players, opponentIndex, map, participants)
```

Comment:
this function feels not great
- why allow empty player input at all?
- why the TBDX stuff?

imo do something like this
(cleaner and kicks the bs)
```lua
function CustomMatchGroupInput._processTeamPlayerMapData(players, opponentIndex, map, participants)
	local numberOfPlayersInMap = 0
	for prefix, player, index in Table.iter.pairsByPrefix(map, 't' .. opponentIndex .. 'p') do
		numberOfPlayersInMap = numberOfPlayersInMap + 1
		local link = mw.ext.TeamLiquidIntegration.resolve_redirect(
			map[prefix .. 'link'] or Variables.varDefault(player .. '_page') or player
		)

		local playerData = {
			played = true,
			cards = CustomMatchGroupInput._readCards(map[prefix .. 'c']),
		}

		local match2playerIndex = CustomMatchGroupInput._fetchMatch2PlayerIndexOfPlayer(players, link)

		if not match2playerIndex then
			table.insert(players, {
				--check the format in the existing players field please and adjust accordingly
				name = link:gsub(' ', '_'),
				displayname = player,
			})
			match2playerIndex = #players
		end

		participants[opponentIndex .. '_' .. match2playerIndex] =  playerData
	end

	return numberOfPlayersInMap
end
```